### PR TITLE
Swift Optimizer: add side effect analysis

### PR DESCRIPTION
**Pull request by eeckstein**
So far, function effects only included escape effects. This PR adds side-effects.
It also involves refactoring of the existing escape effects. The SIL effect syntax changed a bit. Details are in docs/SIL.rst

The new ComputeSideEffects pass computes the side effects for a function, which consists of argument- and global effects.
This is similar to the ComputeEscapeEffects pass, just for side-effects.

Side effects are not used yet. This PR just adds the data structures and the pass which computes the effects.

For details see the commit messages.

### CODE REVIEW

**Comment by eeckstein** [comment]
@swift-ci test

**Comment by eeckstein** [comment]
@swift-ci test

File: SwiftCompilerSources/Sources/SIL/Effects.swift [hunk 1]
```diff
@@ -10,199 +10,507 @@
...
+    }
+
+    public mutating func merge(with other: GlobalEffects) {
+      memory.merge(with: other.memory)
+      ownership.merge(with: other.ownership)
+      allocates = allocates || other.allocates
+    }
+
+    /// Removes effects, which cannot occur for an `argument` value with a given `convention`.
+    public func resticedTo(argument: Value, withConvention convention: ArgumentConvention) -> GlobalEffects {
```
Comment (nate-chandler) [inline SwiftCompilerSources/Sources/SIL/Effects.swift:414]: Is this `restrictedTo`?

Comment (eeckstein) [reply SwiftCompilerSources/Sources/SIL/Effects.swift:414]: yes

**Comment by eeckstein** [comment]
@swift-ci test

**Pull request merged**